apps/application-system/api/src/app/modules/application/charge/application-charge.service.ts

```diff
@@ -17,21 +18,44 @@ export class ApplicationChargeService {
     this.logger = logger.child({ context: 'ApplicationChargeService' })
   }
 
-  async deleteCharge(application: Pick<Application, 'id' | 'externalData'>) {
+  async deleteCharge(
+    application: Pick<Application, 'id' | 'externalData' | 'typeId' | 'state'>,
+  ) {
     try {
       const payment = await this.paymentService.findPaymentByApplicationId(
         application.id,
       )
 
-      // No need to delete charge if already paid or never existed
-      if (!payment || payment.fulfilled) {
+      // No need to delete charge if never existed
+      if (!payment) {
         return
       }
 
+      // No need to delete charge if already paid (and should not be refunded)
+      if (payment.fulfilled) {
+        const template = await getApplicationTemplateByTypeId(
+          application.typeId,
+        )
+
+        const stateConfig =
+          template.stateMachineConfig.states[application.state]
+
+        if (!stateConfig.meta?.lifecycle?.shouldDeleteCharge) {
```

Comment:
And instead of checking external data if id exists check `shouldDeleteCharge` beforehand an then try to delete a charge.